File: src/extras-kcompletion/libkcompletionmatches.cpp
#include <KCompletionMatches>
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <kcompletionmatches.h>
#include "libkcompletionmatches.h"
#include "libkcompletionmatches.hxx"

KCompletionMatches* KCompletionMatches_new(bool sort) {
    return new KCompletionMatches(sort);
}

KCompletionMatches* KCompletionMatches_new2(const KCompletionMatches* param1) {
    return new KCompletionMatches(*param1);
}

void KCompletionMatches_OperatorAssign(KCompletionMatches* self, const KCompletionMatches* param1) {
    self->operator=(*param1);
}

void KCompletionMatches_RemoveDuplicates(KCompletionMatches* self) {
    self->removeDuplicates();
}

libqt_list /* of libqt_string */ KCompletionMatches_List(const KCompletionMatches* self) {
    QList<QString> _ret = self->list();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QString _lv_ret = _ret[i];
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_b = _lv_ret.toUtf8();
        libqt_string _lv_str;
        _lv_str.len = _lv_b.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_b.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

bool KCompletionMatches_Sorting(const KCompletionMatches* self) {
    return self->sorting();
}

libqt_list /* of libqt_string */ KCompletionMatches_List1(const KCompletionMatches* self, bool sort) {
    QList<QString> _ret = self->list(sort);
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QString _lv_ret = _ret[i];
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_b = _lv_ret.toUtf8();
        libqt_string _lv_str;
        _lv_str.len = _lv_b.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_b.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

void KCompletionMatches_Delete(KCompletionMatches* self) {
    delete self;
}

File: src_original/extras-kcompletion/libkcompletionmatches.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://api.kde.org/kcompletionmatches.html
pub const kcompletionmatches = struct {
    /// New constructs a new KCompletionMatches object.
    ///
    /// ``` sort: bool ```
    pub fn New(sort: bool) QtC.KCompletionMatches {
        return qtc.KCompletionMatches_new(sort);
    }

    /// New2 constructs a new KCompletionMatches object.
    ///
    /// ``` param1: QtC.KCompletionMatches ```
    pub fn New2(param1: ?*anyopaque) QtC.KCompletionMatches {
        return qtc.KCompletionMatches_new2(@ptrCast(param1));
    }

    /// [Qt documentation](https://api.kde.org/kcompletionmatches.html#operator-eq)
    ///
    /// ``` self: QtC.KCompletionMatches, param1: QtC.KCompletionMatches ```
    pub fn OperatorAssign(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.KCompletionMatches_OperatorAssign(@ptrCast(self), @ptrCast(param1));
    }

    /// [Qt documentation](https://api.kde.org/kcompletionmatches.html#removeDuplicates)
    ///
    /// ``` self: QtC.KCompletionMatches ```
    pub fn RemoveDuplicates(self: ?*anyopaque) void {
        qtc.KCompletionMatches_RemoveDuplicates(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kcompletionmatches.html#list)
    ///
    /// ``` self: QtC.KCompletionMatches, allocator: std.mem.Allocator ```
    pub fn List(self: ?*anyopaque, allocator: std.mem.Allocator) [][]const u8 {
        const _arr: qtc.libqt_list = qtc.KCompletionMatches_List(@ptrCast(self));
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]const u8, _arr.len) catch @panic("kcompletionmatches.List: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("kcompletionmatches.List: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kcompletionmatches.html#sorting)
    ///
    /// ``` self: QtC.KCompletionMatches ```
    pub fn Sorting(self: ?*anyopaque) bool {
        return qtc.KCompletionMatches_Sorting(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kcompletionmatches.html#list)
    ///
    /// ``` self: QtC.KCompletionMatches, sort: bool, allocator: std.mem.Allocator ```
    pub fn List1(self: ?*anyopaque, sort: bool, allocator: std.mem.Allocator) [][]const u8 {
        const _arr: qtc.libqt_list = qtc.KCompletionMatches_List1(@ptrCast(self), sort);
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]const u8, _arr.len) catch @panic("kcompletionmatches.List1: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("kcompletionmatches.List1: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kcompletionmatches.html#dtor.KCompletionMatches)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.KCompletionMatches ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.KCompletionMatches_Delete(@ptrCast(self));
    }
};
